tests/cpp/test_math_opt.cpp

```diff
@@ -49,4 +49,82 @@ TEST_F(MathOptTest, FastMathTanh) {
   EXPECT_TRUE(ptx_string.find("tanh.approx.f32") != std::string::npos);
 }
 
+TEST_F(MathOptTest, MaxReduceNanProp) {
+  // Check Max NAN behavior for several cases:
+  // 1. No NAN input -> no NAN output
+  // 2. Single NAN input -> NAN output
+  // 3. All NAN input -> NAN output
+
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  auto tv0 = makeSymbolicTensor(1);
+  fusion->addInput(tv0);
+  auto tv1 = max(tv0, {0});
+  fusion->addOutput(tv1);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  auto t0 = at::randn({32}, options);
+
+  KernelExecutor ke;
+  ke.compile(fusion.get(), {t0});
+
+  std::string kernel_code = ke.compiledKernel()->kernelString();
+
+  // No-NAN input
+  auto cg_outputs = ke.run({t0});
+  EXPECT_FALSE(at::any(at::isnan(cg_outputs[0].as<at::Tensor>())).item<bool>());
+
+  // Single NAN input
+  t0[0] = std::numeric_limits<float>::quiet_NaN();
+  cg_outputs = ke.run({t0});
+  EXPECT_TRUE(at::any(at::isnan(cg_outputs[0].as<at::Tensor>())).item<bool>());
+
+  // All NAN input
+  t0 = at::full({32}, std::numeric_limits<float>::quiet_NaN(), options);
+  cg_outputs = ke.run({t0});
+  EXPECT_TRUE(at::any(at::isnan(cg_outputs[0].as<at::Tensor>())).item<bool>());
+}
+
+TEST_F(MathOptTest, FMaxReduceNanProp) {
+  // Check FMax NAN behavior for several cases:
+  // 1. No NAN input -> no NAN output
+  // 2. Single NAN input -> no NAN output
+  // 3. All NAN input -> NAN output
+
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  auto tv0 = makeSymbolicTensor(1);
+  fusion->addInput(tv0);
+  Val* init = ops::binOpIdentity(BinaryOpType::FMax, tv0->dtype());
+  auto tv1 = reductionOp(BinaryOpType::FMax, {0}, init, tv0);
+  fusion->addOutput(tv1);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  auto t0 = at::randn({32}, options);
+
+  KernelExecutor ke;
+  ke.compile(fusion.get(), {t0});
+
+  std::string kernel_code = ke.compiledKernel()->kernelString();
+
+  // Validate that fmax is used by checking the generated code
+  EXPECT_TRUE(kernel_code.find("fmax(") != std::string::npos);
```

Comment:
You can do this the gtest way for better error messages:

```suggestion
  EXPECT_THAT(kernel_code, ::testing::hasSubstr("fmax("));
```